python/pyarrow/array.pyx

```diff
@@ -402,6 +402,31 @@ cdef class BinaryArray(Array):
 
 cdef class DictionaryArray(Array):
 
+    cdef getitem(self, int64_t i):
+        cdef Array dictionary = self.dictionary
+        cdef int64_t index = self.indices[i].as_py()
+        return scalar.box_scalar(dictionary.type, dictionary.sp_array, index)
+
+    property dictionary:
+
+        def __get__(self):
+            cdef CDictionaryArray* darr = <CDictionaryArray*>(self.ap)
```

Comment:
This is solely a cast and does will hopefully be optimised away by the compiler. I'm ok with leaving it here as Cython only allows C89 style variable definitions at the beginning of the function. Leaving it as is, is much more readable.